SilKit/source/core/vasio/RingBuffer.cpp

```diff
@@ -0,0 +1,173 @@
+// SPDX-FileCopyrightText: 2023 Vector Informatik GmbH
+//
+// SPDX-License-Identifier: MIT
+
+#include "RingBuffer.hpp"
+
+#include <algorithm>
+#include <cstring>
+
+#include "silkit/participant/exception.hpp"
+
+namespace SilKit {
+namespace Core {
+
+RingBuffer::RingBuffer(size_t capacity)
+{
+    _buffer.resize(capacity);
+}
+
+void RingBuffer::AdvanceWPos(size_t numBytes)
+{
+    _wPos = (_wPos + numBytes) % Capacity();
+    _size += numBytes;
+
+    SizeCheck();
+}
+
+void RingBuffer::AdvanceRPos(size_t numBytes)
+{
+    _rPos = (_rPos + numBytes) % Capacity();
+    _size -= numBytes;
+
+    SizeCheck();
+}
+
+bool RingBuffer::Peek(std::vector<uint8_t>& elem) const
+{
+    // make sure, we only copy as many bytes as are contained in the buffer
+    if (elem.size() > _size)
+    {
+        return false;
+    }
+
+    // copy data from first contiguous array (of occupied memory)
+    size_t numBytesArrayOne = std::min((Capacity() - _rPos), elem.size());
+    memcpy(elem.data(), _buffer.data() + _rPos, numBytesArrayOne);
+
+    // copy data from second contiguous array (of occupied memory)
+    if (numBytesArrayOne < elem.size())
+    {
+        memcpy(elem.data() + numBytesArrayOne, _buffer.data(), elem.size() - numBytesArrayOne);
+    }
+
+    return true;
+}
+
+bool RingBuffer::Read(std::vector<uint8_t>& elem)
+{
+    if (!Peek(elem))
+    {
+        return false;
+    }
+
+    AdvanceRPos(elem.size());
+
+    return true;
+}
+
+std::vector<MutableBuffer> RingBuffer::GetWritingBuffers()
```

Comment:
To me it seems a bit wasteful to always create a new `vector` in each i/o read (whic allocates once or twice each time this method is called!). I'd prefer to pass the `std::vector<MutableBuffer>` as a `&` parameter to the function (similar to `Peek(...)` and `Read(...)`).

Reply:
Done.